### PR TITLE
Add AppArmor GA fields and validation

### PR DESCRIPTION
**Pull request by mccormickt**
#### What this PR does / why we need it:
This PR adds a GA field for AppArmorProfile for SecurityContext's at the pod and container level.

#### Which issue(s) this PR fixes:

Fixes #24

#### Special notes for your reviewer:
This PR covers API fields added as part of [KEP-24](https://github.com/kubernetes/enhancements/pull/3298).  A separate PR may need to be created to cover certain annotation to field conversions and test cases.

#### Does this PR introduce a user-facing change?

```release-note
AppArmor support has graduated to GA. A new `apparmorProfile` field is added to pod and container securityContext objects. Support for `container.apparmor.security.alpha.kubernetes.io/...` annotations is deprecated, and will be removed in v1.30.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```doc
```

### CODE REVIEW

File: pkg/apis/core/annotation_key_constants.go [hunk 1]
```diff
@@ -34,6 +34,24 @@ const (
 	// in the Annotations of a Node.
 	TaintsAnnotationKey string = "scheduler.alpha.kubernetes.io/taints"
 
+	// AppArmorBetaContainerAnnotationKeyPrefix is the prefix to an annotation key specifying a container's apparmor profile.
```
Comment (sftim) [inline pkg/apis/core/annotation_key_constants.go:37]: (nit)
```suggestion
	// AppArmorBetaContainerAnnotationKeyPrefix is the prefix to an annotation key specifying a container's AppArmor profile.
```

File: pkg/apis/core/annotation_key_constants.go [hunk 2]
```diff
@@ -34,6 +34,24 @@ const (
...
+	AppArmorBetaAllowedProfilesAnnotationKey string = "apparmor.security.beta.kubernetes.io/allowedProfileNames"
+
+	// AppArmorBetaProfileRuntimeDefault is the profile specifying the runtime default.
+	AppArmorBetaProfileRuntimeDefault string = "runtime/default"
+
+	// AppArmorBetaProfileNamePrefix is the prefix for specifying profiles loaded on the node.
+	AppArmorBetaProfileNamePrefix string = "localhost/"
+
+	// AppArmorBetaProfileNameUnconfined is the Unconfined AppArmor profile
+	AppArmorBetaProfileNameUnconfined string = "unconfined"
```
Comment (sftim) [inline pkg/apis/core/annotation_key_constants.go:53]: How come the comment uses title case (“Unconfined”)?

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -5506,6 +5535,12 @@ type SecurityContext struct {
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	SeccompProfile *SeccompProfile
+	// The apparmor options to use by this container. If apparmor options are
```
Comment (sftim) [inline pkg/apis/core/types.go:5538]: (nit)
```suggestion
	// The AppArmor options to use for this container. If AppArmor options are
```

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -5506,6 +5535,12 @@ type SecurityContext struct {
...
+	// provided at both the pod & container level, the container options
+	// override the pod options.
+	// Note that this field cannot be set when spec.os.name is windows.
```
Comment (sftim) [inline pkg/apis/core/types.go:5541]: _aside_
```suggestion
	// Note that this field cannot be set when spec.os.name is "windows".
```
(and change other occurrences)

File: staging/src/k8s.io/api/core/v1/annotation_key_constants.go [hunk 5]
```diff
@@ -55,10 +55,15 @@ const (
 	SeccompLocalhostProfileNamePrefix = "localhost/"
 
 	// AppArmorBetaContainerAnnotationKeyPrefix is the prefix to an annotation key specifying a container's apparmor profile.
```
Comment (sftim) [inline staging/src/k8s.io/api/core/v1/annotation_key_constants.go:57]: (cleanup nit)
```suggestion
	// AppArmorBetaContainerAnnotationKeyPrefix is the prefix to an annotation key specifying a container's AppArmor profile.
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 6]
```diff
@@ -145,6 +148,9 @@ type kubeGenericRuntimeManager struct {
 	// PodState provider instance
 	podStateProvider podStateProvider
 
+	// Use RuntimeDefault as the default seccomp profile for all workloads.
```
Comment (sftim) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:151]:
```suggestion
	// Use the default AppArmor profile for all workloads.
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 7]
```diff
@@ -130,6 +130,9 @@ type kubeGenericRuntimeManager struct {
 	// The directory path for seccomp profiles.
 	seccompProfileRoot string
 
+	// The directory path for apparmor profiles.
```
Comment (sftim) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:133]: (nit)
```suggestion
	// The directory path for AppArmor profiles.
```

File: pkg/apis/core/validation/validation_test.go [hunk 8]
```diff
@@ -9608,6 +9642,26 @@ func TestValidatePod(t *testing.T) {
 				Spec: validPodSpec(nil),
 			},
 		},
+		"AppArmor profile annotation must apply to a container": {
+			expectedError: "metadata.annotations[container.apparmor.security.beta.kubernetes.io/fake-ctr]",
```
Comment (sftim) [inline pkg/apis/core/validation/validation_test.go:9646]: We'd change this again for GA, right?

File: pkg/apis/core/validation/validation.go [hunk 9]
```diff
@@ -4257,6 +4301,72 @@ func validateSeccompAnnotationsAndFieldsMatch(annotationValue string, seccompFie
 	return nil
 }
 
+// ValidateAppArmorAnnotationsAndFields iterates through all containers and ensure that when both appArmorProfile and apparmor annotations exist they match.
```
Comment (sftim) [inline pkg/apis/core/validation/validation.go:4304]:
```suggestion
// ValidateAppArmorAnnotationsAndFields iterates through all containers and ensure that when both appArmorProfile and AppArmor annotations exist they match.
```

File: pkg/apis/core/validation/validation.go [hunk 10]
```diff
@@ -4013,6 +4019,43 @@ func validateSeccompProfileType(fldPath *field.Path, seccompProfileType core.Sec
...
+
+	if ap.Type == core.AppArmorProfileTypeLocalhost {
+		if ap.LocalhostProfile == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("localhostProfile"), "must be set when apparmor type is Localhost"))
+		} else {
+			allErrs = append(allErrs, validateLocalDescendingPath(*ap.LocalhostProfile, fldPath.Child("localhostProfile"))...)
+		}
+	} else {
+		if ap.LocalhostProfile != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("localhostProfile"), ap, "can only be set when apparmor type is Localhost"))
```
Comment (sftim) [inline pkg/apis/core/validation/validation.go:4052]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath.Child("localhostProfile"), ap, "can only be set when AppArmor type is Localhost"))
```

File: pkg/apis/core/validation/validation.go [hunk 11]
```diff
@@ -4013,6 +4019,43 @@ func validateSeccompProfileType(fldPath *field.Path, seccompProfileType core.Sec
...
+		return allErrs
+	}
+
+	if err := validateAppArmorProfileType(fldPath.Child("type"), ap.Type); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if ap.Type == core.AppArmorProfileTypeLocalhost {
+		if ap.LocalhostProfile == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("localhostProfile"), "must be set when apparmor type is Localhost"))
```
Comment (sftim) [inline pkg/apis/core/validation/validation.go:4046]:
```suggestion
			allErrs = append(allErrs, field.Required(fldPath.Child("localhostProfile"), "must be set when AppArmor type is Localhost"))
```

File: pkg/apis/core/types.go [hunk 12]
```diff
@@ -3151,8 +3151,37 @@ type PodSecurityContext struct {
...
+// AppArmorProfileType defines the supported apparmor profile types.
+type AppArmorProfileType string
+
+const (
+	// AppArmorProfileTypeUnconfined is when no apparmor profile is applied (A.K.A. unconfined).
+	AppArmorProfileTypeUnconfined AppArmorProfileType = "Unconfined"
+	// AppArmorProfileTypeRuntimeDefault represents the default container runtime apparmor profile.
+	AppArmorProfileTypeRuntimeDefault AppArmorProfileType = "RuntimeDefault"
+	// AppArmorProfileTypeLocalhost represents custom made profiles stored on the node's disk.
+	AppArmorProfileTypeLocalhost AppArmorProfileType = "Localhost"
```
Comment (sftim) [inline pkg/apis/core/types.go:3182]: How about:
```suggestion
	// AppArmorProfileTypeUnconfined is when no AppArmor profile is applied (also referred to as unconfined).
	AppArmorProfileTypeUnconfined AppArmorProfileType = "Unconfined"
	// AppArmorProfileTypeRuntimeDefault represents the default container runtime AppArmor profile.
	AppArmorProfileTypeRuntimeDefault AppArmorProfileType = "RuntimeDefault"
	// AppArmorProfileTypeLocalhost represents custom made profiles found in the node's filesystem tree.
	AppArmorProfileTypeLocalhost AppArmorProfileType = "Localhost"
```
(Nodes might not have a disk)

File: pkg/apis/core/types.go [hunk 13]
```diff
@@ -3151,8 +3151,37 @@ type PodSecurityContext struct {
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	SeccompProfile *SeccompProfile
+	// The apparmor options to use by the containers in this pod.
+	// Note that this field cannot be set when spec.os.name is windows.
+	// +optional
+	AppArmorProfile *AppArmorProfile
+}
+
+// AppArmorProfile defines a pod/container's apparmor profile settings.
```
Comment (sftim) [inline pkg/apis/core/types.go:3160]:
```suggestion
// AppArmorProfile defines a pod/container's AppArmor profile settings.
```

File: pkg/apis/core/types.go [hunk 14]
```diff
@@ -3151,8 +3151,37 @@ type PodSecurityContext struct {
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	SeccompProfile *SeccompProfile
+	// The apparmor options to use by the containers in this pod.
```
Comment (sftim) [inline pkg/apis/core/types.go:3154]:
```suggestion
	// The AppArmor options to use by the containers in this pod.
```

**Review by sftim — COMMENTED** [review]
Some early feedback. Hope it's useful.

File: pkg/apis/core/annotation_key_constants.go (hunk 2 above)
Comment (jan0ski) [reply pkg/apis/core/annotation_key_constants.go:53]: I think it was due to the struct `type` field being `Unconfined` vs the annotation string `unconfined`, but more or less unintentional.

File: pkg/apis/core/validation/validation_test.go (hunk 8 above)
Comment (jan0ski) [reply pkg/apis/core/validation/validation_test.go:9646]: This looks unneeded, as pod or container fields wouldn't have this problem, where annotations do.

**Comment by sftim** [comment]
:thought_balloon: if someone tries to set one of the deprecated annotations, does the API server respond with a [warning](https://kubernetes.io/blog/2020/09/03/warnings/)?

That might be something to track as a feature-request issue.

**Comment by liggitt** [comment]
It looks like the KEP for this isn't merged yet (https://github.com/kubernetes/enhancements/pull/3298). Once it is, and is slotted into a release, this can be reviewed.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

**Pull request closed without merging**